In [3]:
import pandas as pd

# What factors were the most useful in predicting wage?

In [4]:
# Classifying the wage variable, if wage is above the median value is 1, otherwise 0
# Read the file into a dataframe
df = pd.read_csv('wage_data.txt', sep=r'\s+')

# Calculate the median wage
median_wage = df['wage'].median()

# Classify the wage variable
df['wage_class'] = df['wage'].apply(lambda x: 1 if x > median_wage else 0)
df.head()


education  south  sex  experience  union  wage  age  race  occupation  \
0          8      0    1          21      0  5.10   35     2           6   
1          9      0    1          42      0  4.95   57     3           6   
2         12      0    0           1      0  6.67   19     3           6   
3         12      0    0           4      0  4.00   22     3           6   
4         12      0    0          17      0  7.50   35     3           6   

   sector  marr  wage_class  
0       1     1           0  
1       1     1           0  
2       1     0           0  
3       0     0           0  
4       0     1           0

In [9]:
# create one-hot for occupation
df["occ_1"] = df.occupation.apply(lambda x: 1 if x == 1 else 0)
df["occ_2"] = df.occupation.apply(lambda x: 1 if x == 2 else 0)
df["occ_3"] = df.occupation.apply(lambda x: 1 if x == 3 else 0)
df["occ_4"] = df.occupation.apply(lambda x: 1 if x == 4 else 0)
df["occ_5"] = df.occupation.apply(lambda x: 1 if x == 5 else 0)
df["occ_6"] = df.occupation.apply(lambda x: 1 if x == 6 else 0)

# create one-hot variables for sector
df["sec_0"] = df.sector.apply(lambda x: 1 if x == 0 else 0)
df["sec_1"] = df.sector.apply(lambda x: 1 if x == 1 else 0)
df["sec_2"] = df.sector.apply(lambda x: 1 if x == 2 else 0)

# create one-hot variables for race
df["race_1"] = df.race.apply(lambda x: 1 if x == 1 else 0)
df["race_2"] = df.race.apply(lambda x: 1 if x == 2 else 0)
df["race_3"] = df.race.apply(lambda x: 1 if x == 3 else 0)

In [10]:
X_train, X_test, y_train, y_test = train_test_split(df[['education', 'south', 'sex', 'union', 'race', 'occupation', 'sector', 'marr']], df['wage_class'], test_size=0.2, random_state=42)

# Create an instance of the logistic regression model
model = LogisticRegression()

# Train the model on the training data
model.fit(X_train, y_train)

# Make predictions on the testing data
y_pred = model.predict(X_test)

# Calculate the accuracy of the model
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)


Accuracy: 0.719626168224299


In [12]:
from sklearn.ensemble import RandomForestClassifier

# Create an instance of the random forest classifier
rf_model = RandomForestClassifier()

# Train the model on the training data
rf_model.fit(X_train, y_train)

# Make predictions on the testing data
rf_y_pred = rf_model.predict(X_test)

# Calculate the accuracy of the model
rf_accuracy = accuracy_score(y_test, rf_y_pred)
print("Random Forest Accuracy:", rf_accuracy)


Random Forest Accuracy: 0.7102803738317757


The accuracy of the LogisticRegression model is 0.7196, while the accuracy of the RandomForestClassifier model is 0.7103. This suggests that the LogisticRegression model performs slightly better in terms of accuracy, although the size of the dataset might suggest that this difference is insignificant.

The LogisticRegression model is a linear model that uses a logistic function to model the probability of a binary outcome. On the other hand, the RandomForestClassifier model is an ensemble model that combines multiple decision trees to make predictions. Both models are relatively accuate in their predictions.


In [13]:
coefficients = model.coef_
print("Coefficients:", coefficients)


Coefficients: [[ 0.23131491 -0.48837273 -0.53752226  1.36293902  0.17040296 -0.1524875
   0.66793629  0.66761087]]


Here are the coefficients for each feature:

education: 0.23131491
south: -0.48837273
sex: -0.53752226
union: 1.36293902
race: 0.17040296
occupation: -0.1524875
sector: 0.66793629
marr: 0.66761087

We can conclude that union is the variable with the strongest correlation with wage class.